tsl/src/nodes/decompress_chunk/exec.c

```diff
@@ -291,60 +424,116 @@ decompress_chunk_begin(CustomScanState *node, EState *estate, int eflags)
 		}
 	}
 
-	initialize_column_state(state);
-
 	node->custom_ps = lappend(node->custom_ps, ExecInitNode(compressed_scan, estate, eflags));
-
-	state->per_batch_context = AllocSetContextCreate(CurrentMemoryContext,
-													 "DecompressChunk per_batch",
-													 ALLOCSET_DEFAULT_SIZES);
 }
 
-static void
-initialize_batch(DecompressChunkState *state, TupleTableSlot *compressed_slot,
-				 TupleTableSlot *decompressed_slot)
+void
+decompress_initialize_batch(DecompressChunkState *chunk_state, DecompressBatchState *batch_state,
+							TupleTableSlot *subslot)
 {
 	Datum value;
 	bool isnull;
 	int i;
-	MemoryContext old_context = MemoryContextSwitchTo(state->per_batch_context);
-	MemoryContextReset(state->per_batch_context);
 
-	state->total_batch_rows = 0;
-	state->current_batch_row = 0;
+	Assert(batch_state->initialized == false);
 
-	for (i = 0; i < state->num_columns; i++)
+	/* Batch states can be re-used skip tuple slot creation in that case */
+	if (batch_state->compressed_slot == NULL)
 	{
-		DecompressChunkColumnState *column = &state->columns[i];
+		batch_state->compressed_slot =
+			MakeSingleTupleTableSlot(subslot->tts_tupleDescriptor, subslot->tts_ops);
+	}
+	else
+	{
+		ExecClearTuple(batch_state->compressed_slot);
+	}
+
+	ExecCopySlot(batch_state->compressed_slot, subslot);
```

Comment:
This copies the memory unnecessarily, can we somehow "consume" it without copying? This batch will be the only user of the compressed data. Probably not so important, just wondering.